## Numpy Library Exercises

In [ ]:
import numpy as np
import pandas as pd

In [2]:
from numpy.typing import NDArray
from collections.abc import Hashable, Iterable

**1.** Given a random array, change the sign of elements whose values ​​are between 3 and 8

In [3]:
def change_sing(arr: NDArray[np.number]) -> NDArray[np.number]:
    arr[(arr >= 3) & (arr <= 8)] *= -1
    return arr

ar = np.random.randint(0, 10, 10)
print(ar)
print(change_sing(ar))

[6 7 8 9 5 7 8 3 4 8]
[-6 -7 -8  9 -5 -7 -8 -3 -4 -8]


**2.** Replace the maximum element of a random array with 0

In [4]:
def change_max(arr: NDArray[np.number]) -> NDArray[np.number]:
    #changes ALL the max elements if they repeat
    arr[arr == arr.max()] = 0
    return arr

ar = np.random.randint(0, 10, 10)

print(ar)
print(change_max(ar))

[9 2 2 2 3 3 0 9 3 7]
[0 2 2 2 3 3 0 0 3 7]


**3.** Construct the direct product of arrays (all combinations with each element). A two-dimensional array is provided as input.

In [5]:
def direct_product(arrs: NDArray[np.number]) -> NDArray[np.number]:
    grids = np.meshgrid(*arrs, indexing='ij')

    # stack -> the last axis contains a complete combination & reshape -> flattens the grid so that one row = one combination
    return np.stack(grids, axis=-1).reshape(-1,len(arrs)) 

arr = np.array([[1, 2], [3, 4], [5, 6]])  
result = direct_product(arr)
print(result)

[[1 3 5]
 [1 3 6]
 [1 4 5]
 [1 4 6]
 [2 3 5]
 [2 3 6]
 [2 4 5]
 [2 4 6]]


**4.** Given two arrays, A (8x3) and B (2x2), find rows in A that contain elements from each row in B, regardless of the order of the elements in B.

In [6]:
def find_shared_rows(A: NDArray[np.number], B: NDArray[np.number]) -> NDArray[np.number]:
    # broadcasting -> shape (8, 2, 3, 2)
    comparison = A[:, None, :, None] == B[None, :, None, :]
    contains = comparison.any(axis=(2, 3))   # shape (8, 2) -> for each row A and B
    mask = contains.all(axis=1)   # shape (8,) -> if shared ellement for BOTH B rows
    return A[mask]

A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [1, 5, 9],
    [2, 4, 6],
    [3, 6, 9],
    [1, 4, 7],
    [2, 5, 8]
])

B = np.array([
    [1, 9],
    [4, 8]
])
print(find_shared_rows(A,B))

[[7 8 9]
 [1 4 7]]


**5.** Given a 10x3 matrix, find rows of unequal values ​​(for example, the row [2,2,3] remains, the row [3,3,3] is deleted)

In [7]:
def find_notequal(arr: NDArray[np.number]) -> NDArray[np.number]:
    return arr[~np.all(arr == arr[:,[0]], axis=1)]

arr = np.array([
    [2, 2, 3],
    [3, 3, 3],
    [1, 5, 1],
    [7, 7, 7],
    [4, 4, 9]
])
find_notequal(arr)

array([[2, 2, 3],
       [1, 5, 1],
       [4, 4, 9]])

**6.** Given a two-dimensional array, remove the rows that are repeated.

In [8]:
def remove_doubles(arr: NDArray[np.number]) -> NDArray[np.number]:
    _,indexes = np.unique(arr, axis=0, return_index=True)
    return arr[np.sort(indexes)]

arr = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [1, 2, 3], 
    [7, 8, 9]
])
remove_doubles(arr)

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

______
______

For each of the following problems (1-5) you need to provide 2 implementations – one without using numpy (Consider that where the input or output should be a numpy array, there will simply be lists), and the second one is fully vectorized using numpy (without using Python loops/map/list comprehension).


__Note 1.__ It can be assumed that all specified objects are non-empty (for example, in __task 1__ there are non-zero elements on the diagonal of the matrix).

__Note 2.__ For most tasks, the solution takes no more than 1-2 lines.

___

In [9]:
from numbers import Number

* __Task 1__: Calculate the product of nonzero elements on the diagonal of a rectangular matrix.  
 For example, for X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) answer is 3.

In [10]:
#python
from math import prod

type Matrix = list[list[int]]

def diag_notzero_prod(arr: Matrix) -> int:
    n = min(len(arr), len(arr[0]))
    return prod(arr[i][i] for i in range(n) if arr[i][i] !=0)  

a = [
    [1, 0, 1],
    [2, 0, 2],
    [3, 0, 3],
    [4, 4, 4]
]
print(diag_notzero_prod(a))

3


In [11]:
#numpy
def diag_notzero_prod_np(arr: NDArray[np.number]) -> np.number:
    d = np.diagonal(arr)
    return np.prod(d[d!=0])

a = np.array([
    [1, 0, 1],
    [2, 0, 2],
    [3, 0, 3],
    [4, 4, 4]
])
print(diag_notzero_prod_np(a))

3


* __Task 2__: Two vectors x and y are given. Check if they define the same multiset.  
  For example, for x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) answer is True.

In [12]:
#python
#Counter is better than sorted in terms of performance (O(n)), 
# but requires additional memory usage
from collections import Counter

type MultisetInput[T: Hashable] = Iterable[T]
def same_multiset[T: Hashable](a: MultisetInput[T], b: MultisetInput[T]) -> bool:
    return Counter(a) == Counter(b)

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
print(same_multiset(x,y))

True


In [13]:
#numpy
def same_multiset_np(X: NDArray[np.number], Y: NDArray[np.number]) -> bool:
    return np.array_equal(np.sort(X), np.sort(Y))

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
print(same_multiset_np(x,y))

True


* __Task 3__: Find the maximum element in the vector x among the elements preceded by zero. 
 For example, for x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) answer is 5.

In [14]:
# python
def max_after_zero(arr: list[Number]) -> Number | None:
    maybe = [arr[i] for i in range(1, len(arr)) if arr[i-1] ==0]
    return max(maybe) if maybe else None

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
print(max_after_zero(x))

5


In [15]:
#numpy
def max_after_zero_np(arr: NDArray[np.number]) -> np.number:
    mask = arr[:-1] == 0
    return np.max(arr[1:][mask])

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
print(max_after_zero_np(x))

5


* __Task 4__: Implement run-length encoding. For some vector x, it is necessary to return a tuple of two vectors of the same length. The first one contains numbers and the second one contains how many times they need to be repeated.  
 For example, for x = np.array([2, 2, 2, 3, 3, 3, 5]) answer is (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [16]:
#python
def rle(x: list[int]) -> tuple[list[int], list[int]]:
    values = []
    counts = []
    prev = x[0]
    count = 1
    for value in x[1:]:
        if value == prev:
            count +=1
        else:
            values.append(prev)
            counts.append(count)
            count = 1
            prev = value
    values.append(prev)
    counts.append(count)
    return values, counts
x = [2, 2, 2, 3, 3, 3, 5]
print(rle(x))

([2, 3, 5], [3, 3, 1])


In [17]:
# numpy
type RLEResult = tuple[NDArray[np.number], NDArray[np.int_]]

def rle_np(arr: NDArray[np.number]) -> RLEResult:
    mask = np.concatenate(([True], arr[1:] != arr[:-1]))
    values = arr[mask]
    indexes = np.concatenate((np.flatnonzero(mask),[len(arr)]))
    counts = np.diff(indexes)
    return values, counts

x = np.array([2, 2, 2, 3, 3, 3, 5])
print(rle_np(x))

(array([2, 3, 5]), array([3, 3, 1]))


* __Task 5__: Two samples of objects are given - X and Y. Calculate the matrix of Euclidean distances between objects. Compare with function scipy.spatial.distance.cdist by operating speed.

In [29]:
#python
type Points = list[list[float]]

def cartesian_dist(X: Points, Y: Points) -> Points:
    result = []
    for x in X:
        row = []
        for y in Y:
            dist = (sum((x_i - y_i)**2 for x_i, y_i in zip(x,y)))**0.5
            row.append(round(dist, 3))
        result.append(row)
    return result

X = [
    [1, 2],
    [3, 4],
    [5, 6],
    [7, 8]
]

Y = [
    [2, 1],
    [4, 3],
    [6, 5]
]
print(cartesian_dist(X,Y))

[[1.414, 3.162, 5.831], [3.162, 1.414, 3.162], [5.831, 3.162, 1.414], [8.602, 5.831, 3.162]]


In [30]:
#numpy
def cartesian_dist_np(X: NDArray[np.number], Y: NDArray[np.number]) -> NDArray[np.floating]:
    X_sq = np.sum(X**2, axis=1)[:, None]
    Y_sq = np.sum(Y**2, axis=1)[None,:]
    cross = np.dot(X, Y.T)
    res = X_sq + Y_sq - 2*cross
    return np.round(np.sqrt(np.maximum(res, 0)), 3)

X = np.array([
    [1, 2],
    [3, 4],
    [5, 6],
    [7, 8]
])

Y = np.array([
    [2, 1],
    [4, 3],
    [6, 5]
])
print(cartesian_dist_np(X,Y))

[[1.414 3.162 5.831]
 [3.162 1.414 3.162]
 [5.831 3.162 1.414]
 [8.602 5.831 3.162]]


_______
________

In [31]:
#comparison
from scipy.spatial.distance import cdist
import timeit

np.random.seed(42)
X = np.random.rand(1000, 10)
Y = np.random.rand(500, 10)

#repetition num
number = 5

python_time = timeit.timeit(
    lambda: cartesian_dist(X, Y),
    number=number
)

numpy_time = timeit.timeit(
    lambda: cartesian_dist_np(X, Y),
    number=number
)

scipy_time = timeit.timeit(
    lambda: cdist(X, Y),
    number=number
)


print(f"Python: {python_time / number:.3f} сек.")
print(f"NumPy:  {numpy_time / number:.3f} сек.")
print(f"SciPy:  {scipy_time / number:.3f} сек.")

Python: 5.803 сек.
NumPy:  0.019 сек.
SciPy:  0.005 сек.


* #### __Task 6__: CrunchieMunchies __*__

You work in the marketing department of a food company called MyCrunch, which is developing a new type of delicious, healthy cereal called **CrunchieMunchies**.

You want to show consumers how healthy your cereal is compared to other leading brands, so you've compiled nutrition data from several different competitors.

Your task is to use Numpy calculations to analyze this data and prove that your **СrunchieMunchies** - the healthiest choice for consumers.


In [21]:
import numpy as np

1. View the cereal.csv file. This file contains calorie counts for various brands of cereal. Download the data from the file and save it as calorie_stats.

In [22]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. One serving of CrunchieMunchies contains 60 calories. How much higher is the average calorie count of your competitors?

Store the answer in the variable average_calories and print the variable in the terminal

In [ ]:
average_calories = np.mean(calorie_stats) - 60
print(f"The average calorie count of the competitors is {average_calories:.3f} higher than ours")

The average calorie count of the competitors is 46.883


3. Does the average calorie count accurately reflect the distribution of the data set? Let's sort the data and see.

Sort the data and store the result in the calorie_stats_sorted variable. Print the sorted information.

In [24]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. It looks like most values ​​are above the mean. Let's see if the median is the most accurate measure of the data set.

Calculate the median of the data set and save your answer in median_calories. Print the median so you can see how it compares to the mean.

In [25]:
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. While the median shows that at least half of our values ​​are over 100 calories, it would be more impressive to show that a significant portion of the competitors have higher calorie counts than CrunchieMunchies.

Calculate various percentiles and print them until you find the lowest percentile that exceeds 60 calories. Store this value in the nth_percentile variable.

In [60]:
percentiles = np.arange(1, 101)
values = np.percentile(calorie_stats, percentiles)

cutoff_percentile = np.argmax(values > 60)

idx = [f"{p}%" for p in percentiles[:cutoff_percentile+1]]
print(pd.Series(values[:cutoff_percentile+1], index=idx).round(3))
nth_percentile = percentiles[cutoff_percentile]
print(f'{nth_percentile}%')

1%    50.0
2%    50.0
3%    55.6
4%    70.0
dtype: float64
4%


6. While the percentile shows us that most competitors have much higher calorie counts, it's an awkward concept to use in marketing materials.

Instead, let's calculate the percentage of cereals that contain more than 60 calories per serving. Save your answer in the more_calories variable and print it.

In [39]:
more_calories = np.mean(calorie_stats > 60)*100
print(f"{more_calories:.3f} %")

96.104 %


7. This is a really high percentage. This will be very useful as we promote CrunchieMunchies. But one question is: how big are the differences in the dataset? Can we generalize that most cereals contain around 100 calories or is the range even wider?

Calculate the magnitude of the deviation by finding the standard deviation. Save your answer in calorie_std and print it on the terminal. How can we incorporate this value into our analysis?

In [40]:
calorie_std = np.std(calorie_stats)
print(f"{calorie_std:.3f}")

19.357


While the average calorie count across competitor cereals is approximately 107 calories, a standard deviation of roughly 19.38 indicates moderate (it's a relative estimation, it may depend on the level of sensitivity and accuracy) variability in the market. This means it would be inaccurate to generalize that most cereals contain around 100 calories. Competitor products vary in calorie amount, so representing the market with a single average figure might be misleading

8. Please write a short paragraph outlining your findings and how you believe this data can be used to Mycrunch's advantage in marketing CrunchieMunchies.

The analysis of competing cereal brands shows an average calorie count of approximately 107 calories per serving, with a standard deviation of about 19.4 indicating moderate but real variability across the market. The median aligns with the mean, confirming that this average actually reflects a typical product rather than being skewed by extreme values. CrunchieMunchies with its 60 calories per serving has a lower calorie content than approximately 96% of its competitors. This gives the company a competitive advantage for lower-calorie cereal options. MyCrunch should leverage this as a core competitive advantage in marketing.
